### PR TITLE
Consider incident angle when calculating intensity

### PR DESCRIPTION
**Pull request by msz-rai**
This PR:
- takes into account the influence of the incident angle on the intensity value
- adds API call for setting default intensity.
  - intensity considered when hitting entities with no intensity texture set

### CODE REVIEW

File: src/gpu/RaytraceRequestContext.hpp [hunk 1]
```diff
@@ -28,6 +28,8 @@ struct RaytraceRequestContext
 	float nearNonHitDistance;
 	float farNonHitDistance;
 
+	float defaultIntensity;
```
Comment (PawelLiberadzki) [inline src/gpu/RaytraceRequestContext.hpp:31]: Do we agree on default here being 0.0f?

Comment (scorp18pl) [reply src/gpu/RaytraceRequestContext.hpp:31]: Wouldn't 0.5f make more sense?

Comment (alek-kam-robotec-ai) [reply src/gpu/RaytraceRequestContext.hpp:31]: Wouldn't 0.5f make more sense?

**Review by PawelLiberadzki — APPROVED** [review]

File: src/gpu/optixPrograms.cu [hunk 2]
```diff
@@ -203,6 +203,7 @@ extern "C" __global__ void __closesthit__()
 
 		intensity = tex2D<TextureTexelFormat>(entityData.texture, uv[0], uv[1]);
 	}
+	intensity *= cosf(incidentAngle);
```
Comment (scorp18pl) [inline src/gpu/optixPrograms.cu:206]: @msz-rai  Isn't this redundant? The absolute value of the dot product is already known, no need to calculate cos on an acos output. The above incident angle calculation has to be simply split into two statements. One to calculate and store the absolute value of the dot product (this value can be used here). Second one for calculation of the incident angle. To the best of my knowledge tryg functions are rather expensive (i would guess that even more so on the gpu).

File: src/gpu/optixPrograms.cu [hunk 3]
```diff
@@ -188,7 +188,7 @@ extern "C" __global__ void __closesthit__()
 	const Vec3f wNormal = wAB.cross(wCA).normalized();
 	const float incidentAngle = acosf(fabs(wNormal.dot(rayDir)));
```
Comment (alek-kam-robotec-ai) [inline src/gpu/optixPrograms.cu:189]: The line I'm suggesting to split.

File: src/gpu/RaytraceRequestContext.hpp (hunk 1 above)
Comment (msz-rai) [reply src/gpu/RaytraceRequestContext.hpp:31]: Discussed offline

File: src/gpu/optixPrograms.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/gpu/optixPrograms.cu:206]: Good point. :+1:

**Pull request merged**